File: 3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s == "")
            return 0;
        std::vector<int> f(s.size());
        f[0] = 1;
        for(int i = 1;i < s.size();i++){
            int c = -1;
            for(int j = i-1;j >= i-f[i-1];j--){
                if(s[i] == s[j]){
                    c = j;
                    break;
                }
            }
            if(c >= 0)
                f[i] = i-c;
            else
                f[i] = f[i-1]+1;
        }
        return *std::max_element(f.begin(), f.end());
    }
};

File: 3.无重复字符的最长子串.rs
/*
 * @lc app=leetcode.cn id=3 lang=rust
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
use std::collections::HashSet;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        if s.len() <= 1 {
            return s.len() as i32;
        }
    
        let mut head = 0;
        let mut tail = 1;
        let bytes = s.as_bytes();
        let mut chars_collection = HashSet::new();
        chars_collection.insert(bytes[0]);
        let mut result = 1;
        while tail < bytes.len() {
            if !chars_collection.contains(&bytes[tail]) {
                chars_collection.insert(bytes[tail]);
                result = result.max(chars_collection.len());
                tail += 1;
            } else {
                chars_collection.remove(&bytes[head]);
                head += 1;
            }
        }
        return result as i32;
    }
}
// @lc code=end

